# nf-core/yascp: Output

## Introduction

This documentation provides a comprehensive overview of the outputs generated by the pipeline.

The pipeline will create the following files in your working directory:

```console
work            # Nextflow working directory
results         # Final results (configurable, see below)
.nextflow_log   # Log file from Nextflow
# Other Nextflow hidden files, eg. history of pipeline runs and old logs.
```

The structure of the results folder is outlined below, providing a snapshot of the diverse outputs from various stages of the pipeline:

- **[preprocessing](#preprocessing)**: Contains data split by modality and preprocessing outputs
- **[doublet_detection](#doublet_detection)**: Includes results identifying artificial doublet cells.
- **[deconvolution](#deconvolution)**: Contains outputs from disentangling mixtures of cells from different donors.
- **[celltype_assignemt](#celltype_assignemt)**: Houses results of cell classification into specific types.
- **[clustering_and_integration](#clustering_and_integration)**: Contains clustered cells grouped by similarities and integrated datasets.
- **[citeseq](#citeseq)**: Includes CITE-seq data, linking transcriptomic and proteomic profiles.
- **[handover](#handover)**: Contains the final processed results.
- **[pipeline_info]**: Stores metadata and logs detailing the pipeline execution process. ðŸ”´**CHECK AND CHANGE IF NEEDED**
- **[yascp_inputs]**: ðŸ”´**THE DESCRIPTION IS NEEDED**

Detailed explanations of each output folder and their corresponding steps are provided below:

## preprocessing

The `preprocessing` folder contains the following subdirectories:

- **data_modalities_split**  
   This folder contains raw and filtered quantification data separated by modality (CITE-seq or hashtag).

- **resources**  
   This folder contains genotype assembly and input files for the test dataset.

- **subset_genotypes**  
   This folder is created only if VCF IDs are specified in the input TSV file. It contains genotypes divided from the pools.

<details>
<summary>Click to view detailed folder structure</summary>

```
preprocessing/
â”œâ”€â”€ data_modalities_split
â”‚Â Â  â”œâ”€â”€ filterd
â”‚Â Â  â”‚Â Â  â””â”€â”€ Pool1
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Gene_Expression-Pool1.h5ad
â”‚Â Â  â”‚Â Â      â””â”€â”€ Pool1__Gene_Expression
â”‚Â Â  â”œâ”€â”€ preprocess
â”‚Â Â  â”‚Â Â  â””â”€â”€ Pool1
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Gene_Expression-Pool1.h5ad
â”‚Â Â  â”‚Â Â      â””â”€â”€ Pool1__Gene_Expression
â”‚Â Â  â””â”€â”€ raw
â”‚Â Â      â””â”€â”€ Pool1
â”‚Â Â          â”œâ”€â”€ Gene_Expression-Pool1.h5ad
â”‚Â Â          â””â”€â”€ Pool1__Gene_Expression
â”œâ”€â”€ recourses
â”‚Â Â  â”œâ”€â”€ Done.tmp
â”‚Â Â  â”œâ”€â”€ full_test_dataset
â”‚Â Â  â”œâ”€â”€ input_test_data_file.tsv
â”‚Â Â  â””â”€â”€ input_test_vcf_file.tsv
â””â”€â”€ subset_genotypes
    â”œâ”€â”€ Genotype___AllExpectedGT_Pool1
    â””â”€â”€ Genotypes_all_pools.tsv

```
</details>

## doublet_detection

The `doublet_detection` folder contains the following subdirectories:

- **DoubletFinder, DoubletDecon, scDblFinder, SCDS**  
   Each folder contains TSV files for each pool. These files include barcodes and labels indicating whether each cell is a singlet or a doublet.

- **scrublet**  
   This folder contains TSV files for each pool with barcodes and labels indicating whether each cell is a multiplet or not. Additionally, this folder includes a subdirectory with plots.

- **doublet_results_combined**  
   This folder contains TSV files for each pool with barcodes and combined labels from multiple tools: Scrublet, SCDS, scDblFinder, DoubletDecon, and DoubletFinder.

- **droplet_type_distribution**  
   This folder contains PNG files with graphs visualizing the distribution of droplet types.

<details>
<summary>Click to view detailed folder structure</summary>

```
doublet_detection
â”œâ”€â”€ DoubletDecon
â”‚Â Â  â””â”€â”€ Pool1__DoubletDecon_doublets_singlets.tsv
â”œâ”€â”€ DoubletFinder
â”‚Â Â  â””â”€â”€ Pool1__DoubletFinder_doublets_singlets.tsv
â”œâ”€â”€ doublet_results_combined
â”‚Â Â  â””â”€â”€ Pool1__doublet_results_combined.tsv
â”œâ”€â”€ droplet_type_distribution
â”‚Â Â  â””â”€â”€ Pool1__droplet_type_distribution.png
â”œâ”€â”€ scDblFinder
â”‚Â Â  â””â”€â”€ Pool1__scDblFinder_doublets_singlets.tsv
â”œâ”€â”€ SCDS
â”‚Â Â  â””â”€â”€ Pool1__scds_doublets_singlets.tsv
â””â”€â”€ scrublet
    â”œâ”€â”€ plots
    â”‚Â Â  â”œâ”€â”€ Pool1boxplot_total_umi_counts.png
    â”‚Â Â  â”œâ”€â”€ Pool1histogram_multiplet_scores_log.png
    â”‚Â Â  â”œâ”€â”€ Pool1histogram_multiplet_scores.png
    â”‚Â Â  â””â”€â”€ Pool1histogram_multiplet_zscores.png
    â””â”€â”€ Pool1scrublet.tsv
```
</details>

## deconvolution
The `deconvolution` folder contains the following subdirectories:

- **vireo_raw**  
   This folder contains genotypes if run in genotype-aware mode
   <details>
   <summary>Click to view detailed folder structure</summary>

   ```
   vireo_raw
   â”œâ”€â”€ correlations.png
   â”œâ”€â”€ Pool1
   â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
   â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
   â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
   â”‚Â Â  â””â”€â”€ vireo_Pool1
   â”œâ”€â”€ donor_corelations_matrix.tsv
   â””â”€â”€ matched_donors.txt

   ```
   </details>

- **vireo_processed**  
   The `vireo_processed` folder contains genotypes renamed to imitate genotype absent mode to ensure consistency in the downstream pipeline tasks
   <details>
   <summary>Click to view detailed folder structure</summary>

   ```
   vireo_processed
   â”œâ”€â”€ assignments_all_pools.tsv
   â””â”€â”€ Pool1
       â”œâ”€â”€ GT_replace_donor_ids_false.tsv
       â”œâ”€â”€ GT_replace_GT_donors.vireo_false.vcf.gz
       â”œâ”€â”€ GT_replace_Pool1_assignments_false.tsv
       â”œâ”€â”€ GT_replace_Pool1__exp.sample_summary_false.txt
       â””â”€â”€ GT_replace_Pool1.sample_summary_false.txt
   ```
   </details>

- **vireo_sub**  
   The `vireo_sub` folder contains vireo permutations to ensure stability in cell assignment
   <details>
   <summary>Click to view detailed folder structure</summary>

   ```
   vireo_sub
   â””â”€â”€ Pool1
    Â Â  â”œâ”€â”€ vireo_____1
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
    Â Â  â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
    Â Â  â”‚Â Â  â””â”€â”€ vireo_Pool1___1
    Â Â  â”œâ”€â”€ vireo_____10
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
    Â Â  â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
    Â Â  â”‚Â Â  â””â”€â”€ vireo_Pool1___10
    Â Â  â”œâ”€â”€ vireo_____2
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
    Â Â  â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
    Â Â  â”‚Â Â  â””â”€â”€ vireo_Pool1___2
    Â Â  â”œâ”€â”€ vireo_____3
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
    Â Â  â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
    Â Â  â”‚Â Â  â””â”€â”€ vireo_Pool1___3
    Â Â  â”œâ”€â”€ vireo_____4
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
    Â Â  â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
    Â Â  â”‚Â Â  â””â”€â”€ vireo_Pool1___4
    Â Â  â”œâ”€â”€ vireo_____5
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
    Â Â  â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
    Â Â  â”‚Â Â  â””â”€â”€ vireo_Pool1___5
    Â Â  â”œâ”€â”€ vireo_____6
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
    Â Â  â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
    Â Â  â”‚Â Â  â””â”€â”€ vireo_Pool1___6
    Â Â  â”œâ”€â”€ vireo_____7
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
    Â Â  â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
    Â Â  â”‚Â Â  â””â”€â”€ vireo_Pool1___7
    Â Â  â”œâ”€â”€ vireo_____8
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
    Â Â  â”‚Â Â  â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
    Â Â  â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
    Â Â  â”‚Â Â  â””â”€â”€ vireo_Pool1___8
    Â Â  â””â”€â”€ vireo_____9
    Â Â      â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
    Â Â      â”œâ”€â”€ dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
    Â Â      â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
    Â Â      â””â”€â”€ vireo_Pool1___9
   ```
   </details>



- **infered_genotypes**  
   The `infered_genotypes` folder contains genotypes called from single-cell data (per donor in a pool)
   <details>
   <summary>Click to view detailed folder structure</summary>

   ```
   infered_genotypes
   â””â”€â”€ Pool1
    Â Â  â”œâ”€â”€ Pool1_headfix_vireo.vcf.gz
    Â Â  â””â”€â”€ Pool1_headfix_vireo.vcf.gz.tbi
   ```
   </details>


- **split_donor_h5ad**  
   The `split_donor_h5ad` folder contains per donor quantification matrix and additional metadata
   <details>
   <summary>Click to view detailed folder structure</summary>

   ```
   split_donor_h5ad
   â””â”€â”€ Pool1
       â”œâ”€â”€ cell_belongings.tsv
       â”œâ”€â”€ donor_level_anndata
       â”‚Â Â  â”œâ”€â”€ donor0.Pool1.barcodes.tsv
       â”‚Â Â  â”œâ”€â”€ donor0.Pool1.h5ad
       â”‚Â Â  â”œâ”€â”€ donor1.Pool1.barcodes.tsv
       â”‚Â Â  â”œâ”€â”€ donor1.Pool1.h5ad
       â”‚Â Â  â”œâ”€â”€ donor2.Pool1.barcodes.tsv
       â”‚Â Â  â”œâ”€â”€ donor2.Pool1.h5ad
       â”‚Â Â  â”œâ”€â”€ doublet.Pool1.barcodes.tsv
       â”‚Â Â  â”œâ”€â”€ doublet.Pool1.h5ad
       â”‚Â Â  â”œâ”€â”€ unassigned.Pool1.barcodes.tsv
       â”‚Â Â  â””â”€â”€ unassigned.Pool1.h5ad
       â”œâ”€â”€ Pool1.donors.h5ad.assigned.tsv
       â”œâ”€â”€ Pool1__donors.h5ad.assigned.tsv
       â”œâ”€â”€ Pool1.donors.h5ad.tsv
       â”œâ”€â”€ Pool1__donors.h5ad.tsv
       â”œâ”€â”€ Pool1_exp__donor_n_cells.tsv
       â”œâ”€â”€ Pool1.h5ad.tsv
       â”œâ”€â”€ vireo_annot.Pool1.h5ad
       â””â”€â”€ Vireo_plots.pdf
   ```
   </details>

- **cellsnp**  
   The `cellsnp` folder contains genotypes called from single-cell data per (per droplet/cell)
   <details>
   <summary>Click to view detailed folder structure</summary>

   ```
   cellsnp
   â””â”€â”€ cellsnp_Pool1
       â”œâ”€â”€ cellSNP.base.vcf.gz
       â”œâ”€â”€ cellSNP.cells.vcf.gz
       â”œâ”€â”€ cellSNP.samples.tsv
       â”œâ”€â”€ cellSNP.tag.AD.mtx
       â”œâ”€â”€ cellSNP.tag.DP.mtx
       â””â”€â”€ cellSNP.tag.OTH.mtx
   ```
   </details>

- **concordances**  
   The `concordances` contains statistics describing how confident the cell-inferred genotype aligns with the reference genotypes.
   <details>
   <summary>Click to view detailed folder structure</summary>

   ```
   concordances
   â”œâ”€â”€ all_variants_description.tsv
   â”œâ”€â”€ becoming_different_donor.png
   â”œâ”€â”€ becoming_doublet_donor.png
   â”œâ”€â”€ becoming_unassigned_donor.png
   â”œâ”€â”€ Pool1
   â”‚Â Â  â”œâ”€â”€ 1090095_1090095-donor3--each_cells_comparison_with_other_donor.tsv
   â”‚Â Â  â”œâ”€â”€ 1709635_1709635-donor5--each_cells_comparison_with_other_donor.tsv
   â”‚Â Â  â”œâ”€â”€ 2288590_2288590-donor6--each_cells_comparison_with_other_donor.tsv
   â”‚Â Â  â”œâ”€â”€ 2743244_2743244-donor7--each_cells_comparison_with_other_donor.tsv
   â”‚Â Â  â”œâ”€â”€ 2768849_2768849-donor4--each_cells_comparison_with_other_donor.tsv
   â”‚Â Â  â”œâ”€â”€ 2998395_2998395-donor2--each_cells_comparison_with_other_donor.tsv
   â”‚Â Â  â”œâ”€â”€ 3183427_3183427-donor0--each_cells_comparison_with_other_donor.tsv
   â”‚Â Â  â”œâ”€â”€ 3699286_3699286-donor1--each_cells_comparison_with_other_donor.tsv
   â”‚Â Â  â”œâ”€â”€ 4853673_4853673-donor9--each_cells_comparison_with_other_donor.tsv
   â”‚Â Â  â”œâ”€â”€ 5154993_5154993-donor8--each_cells_comparison_with_other_donor.tsv
   â”‚Â Â  â”œâ”€â”€ becoming_different_donor.png
   â”‚Â Â  â”œâ”€â”€ becoming_doublet_donor.png
   â”‚Â Â  â”œâ”€â”€ becoming_unassigned_donor.png
   â”‚Â Â  â”œâ”€â”€ cell_belongings.tsv
   â”‚Â Â  â”œâ”€â”€ cellSNP.cells.vcf.gz
   â”‚Â Â  â”œâ”€â”€ Pool1__joined_df_for_plots.tsv
   â”‚Â Â  â”œâ”€â”€ Pool1_subsampling_donor_swap_quantification.tsv
   â”‚Â Â  â”œâ”€â”€ Discordant_reads_becoming_different_donor_no0.png
   â”‚Â Â  â”œâ”€â”€ Discordant_reads_becoming_different_donor.png
   â”‚Â Â  â”œâ”€â”€ Discordant_reads_by_n_sites_becoming_different_donor_no0.png
   â”‚Â Â  â”œâ”€â”€ Discordant_reads_by_n_sites_becoming_different_donor.png
   â”‚Â Â  â”œâ”€â”€ discordant_sites_in_other_donors_noA2G.tsv
   â”‚Â Â  â”œâ”€â”€ Nr_discordant_uninformative_becoming_different_donor.png
   â”‚Â Â  â”œâ”€â”€ sites_becoming_different_donor_no0.png
   â”‚Â Â  â”œâ”€â”€ sites_becoming_different_donor.png
   â”‚Â Â  â”œâ”€â”€ sites_becoming_different_donor_probs.png
   â”‚Â Â  â”œâ”€â”€ sites_becoming_doublet_donor.png
   â”‚Â Â  â”œâ”€â”€ sites_becoming_unassigned_donor.png
   â”‚Â Â  â”œâ”€â”€ sites_vs_concordance.png
   â”‚Â Â  â”œâ”€â”€ stats_Pool1_gt_donor_assignments.csv
   â”‚Â Â  â”œâ”€â”€ sub_Pool1_Expected.vcf.gz
   â”‚Â Â  â”œâ”€â”€ sub_Pool1_GT_Matched.vcf.gz
   â”‚Â Â  â”œâ”€â”€ subplot_sites_vs_concordance.png
   â”‚Â Â  â””â”€â”€ Total_reads_becoming_different_donor.png
   â”œâ”€â”€ Discordant_reads_becoming_different_donor_no0.png
   â”œâ”€â”€ Discordant_reads_becoming_different_donor.png
   â”œâ”€â”€ Discordant_reads_by_n_sites_becoming_different_donor_no0.png
   â”œâ”€â”€ Discordant_reads_by_n_sites_becoming_different_donor.png
   â”œâ”€â”€ joined_df_for_plots.tsv
   â”œâ”€â”€ Nr_discordant_uninformative_becoming_different_donor.png
   â”œâ”€â”€ sites_becoming_different_donor_no0.png
   â”œâ”€â”€ sites_becoming_different_donor.png
   â”œâ”€â”€ sites_becoming_different_donor_probs.png
   â”œâ”€â”€ sites_becoming_doublet_donor.png
   â”œâ”€â”€ sites_becoming_unassigned_donor.png
   â”œâ”€â”€ sites_vs_concordance.png
   â”œâ”€â”€ subplot_sites_vs_concordance.png
   â””â”€â”€ Total_reads_becoming_different_donor.png

   ```
   </details>

- **gtmatch**  
   If genotypes are provided it contains the results of donors assigned by gtcheck.
   <details>
   <summary>Click to view detailed folder structure</summary>

   ```
   gtmatch/
   â”œâ”€â”€ assignments_all_pools.tsv
   â””â”€â”€ Pool1
       â”œâ”€â”€ Done.tmp
       â”œâ”€â”€ Expected_Withing_expected_Pool1.genome
       â”œâ”€â”€ GT_replace_PiHAT_Stats_File_Pool1.csv
       â”œâ”€â”€ InferedExpected_Expected_Infered_Pool1.genome
       â”œâ”€â”€ InferedGTMatched_Expected_Infered_Pool1.genome
       â”œâ”€â”€ InferedOnly_Withing_pool_Pool1.genome
       â”œâ”€â”€ PiHAT_Stats_File_Pool1.csv
       â”œâ”€â”€ Pool1_gt_donor_assignments.csv
       â”œâ”€â”€ pool_Pool1_panel_Pool1_Onek1K_gtcheck_donor_assignments.csv
       â”œâ”€â”€ pool_Pool1_panel_Pool1_Onek1K_gtcheck_score_table.csv
       â””â”€â”€ stats_Pool1_gt_donor_assignments.csv
   ```
   </details>

## celltype_assignemt
The `celltype_assignemt` folder contains the following subdirectories and files:

- **All_Celltype_Assignments.tsv**  
   A combined file with results per barcode/droplet/cell.

- **donor_celltype_report.tsv**  
   Summarized cell counts per donor.

- **tranche_celltype_report.tsv**  
   Summarized cell counts per tranche.

- **scored**  
   This folder contains the results of scored
  
- **azimuth**  
   This folder contains the results of azimuth
  
- **celltypist**  
   This folder contains the results of celltypist

<details>
<summary>Click to view detailed folder structure</summary>

```
celltype_assignemt/
â”œâ”€â”€ All_Celltype_Assignments.tsv
â”œâ”€â”€ azimuth
â”‚Â Â  â””â”€â”€ PBMC
â”‚Â Â      â”œâ”€â”€ AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.mapping_score_umap.pdf
â”‚Â Â      â”œâ”€â”€ AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.mapping_score_vln.pdf
â”‚Â Â      â”œâ”€â”€ AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.ncells_by_type_barplot.pdf
â”‚Â Â      â”œâ”€â”€ AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.prediction_score_umap.pdf
â”‚Â Â      â”œâ”€â”€ AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.prediction_score_vln.pdf
â”‚Â Â      â”œâ”€â”€ AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.query_umap.pdf
â”‚Â Â      â””â”€â”€ AZ_1.pre_QC_adata_Pool1_Pool1_predicted_celltype_l1.tsv
â”œâ”€â”€ celltypist
â”‚Â Â  â”œâ”€â”€ COVID19_Immune_Landscape
â”‚Â Â  â”‚Â Â  â””â”€â”€ Pool1
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Pool1___COVID19_Immune_Landscape___decision_matrix.csv
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Pool1___COVID19_Immune_Landscape___predicted_labels.csv
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Pool1___COVID19_Immune_Landscape___probability_matrix.csv
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Pool1_majority_voting.pdf
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Pool1_over_clustering.pdf
â”‚Â Â  â”‚Â Â      â””â”€â”€ Pool1_predicted_labels.pdf
â”‚Â Â  â”œâ”€â”€ Immune_All_High
â”‚Â Â  â”‚Â Â  â””â”€â”€ Pool1
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Pool1___Immune_All_High___decision_matrix.csv
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Pool1___Immune_All_High___predicted_labels.csv
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Pool1___Immune_All_High___probability_matrix.csv
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Pool1_majority_voting.pdf
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ Pool1_over_clustering.pdf
â”‚Â Â  â”‚Â Â      â””â”€â”€ Pool1_predicted_labels.pdf
â”‚Â Â  â””â”€â”€ Immune_All_Low
â”‚Â Â      â””â”€â”€ Pool1
â”‚Â Â          â”œâ”€â”€ Pool1___Immune_All_Low___decision_matrix.csv
â”‚Â Â          â”œâ”€â”€ Pool1___Immune_All_Low___predicted_labels.csv
â”‚Â Â          â”œâ”€â”€ Pool1___Immune_All_Low___probability_matrix.csv
â”‚Â Â          â”œâ”€â”€ Pool1_majority_voting.pdf
â”‚Â Â          â”œâ”€â”€ Pool1_over_clustering.pdf
â”‚Â Â          â””â”€â”€ Pool1_predicted_labels.pdf
â”œâ”€â”€ donor_celltype_report.tsv
â”œâ”€â”€ scpred
â”‚Â Â  â”œâ”€â”€ AZ_1.pre_QC_adata_Pool1_AZ_1.pre_QC_adata_Pool1__scpred_prediction.tsv
â”‚Â Â  â””â”€â”€ AZ_1.pre_QC_adata_Pool1_hier_scpred.RDS
â””â”€â”€ tranche_celltype_report.tsv
```
</details>

## clustering_and_integration

The `clustering_and_integration` folder contains integrated and clustered data, along with statistics and plots that describe the performance of the integration and clustering processes.

<details>
<summary>Click to view detailed folder structure</summary>

```
clustering_and_integration/
â”œâ”€â”€ normalize=total_count.vars_to_regress=none
â”‚Â Â  â”œâ”€â”€ adatametadata.tsv.gz
â”‚Â Â  â”œâ”€â”€ adatanormalized.h5ad
â”‚Â Â  â”œâ”€â”€ adatanormalized_pcacounts.h5ad
â”‚Â Â  â”œâ”€â”€ adatanormalized_pca.h5ad
â”‚Â Â  â”œâ”€â”€ adatanormalized_pcaknee.tsv
â”‚Â Â  â”œâ”€â”€ adatapcs.tsv.gz
â”‚Â Â  â”œâ”€â”€ donor_level_anndata_QCfiltered
â”‚Â Â  â”‚Â Â  â””â”€â”€ Pool1___sample_QCd_adata.h5ad
â”‚Â Â  â”œâ”€â”€ plots
â”‚Â Â  â”œâ”€â”€ reduced_dims-null-bbknn.batch=experiment_id.n_pcs=20
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ cluster.number_neighbors=-1.method=leiden.resolution=0.1
â”‚Â Â  â”‚Â Â  â”‚Â Â  â”œâ”€â”€ clustering_0.1clustered.h5ad
â”‚Â Â  â”‚Â Â  â”‚Â Â  â”œâ”€â”€ clustering_0.1clustered.tsv.gz
â”‚Â Â  â”‚Â Â  â”‚Â Â  â”œâ”€â”€ dotplot_sampleclustering_0.1clustered_ncells0.pdf
â”‚Â Â  â”‚Â Â  â”‚Â Â  â”œâ”€â”€ dotplot_sampleclustering_0.1clustered_ncellsless5.pdf
â”‚Â Â  â”‚Â Â  â”‚Â Â  â”œâ”€â”€ dotplot_sampleclustering_0.1clustered.pdf
â”‚Â Â  â”‚Â Â  â”‚Â Â  â”œâ”€â”€ plots
â”‚Â Â  â”‚Â Â  â”‚Â Â  â”œâ”€â”€ sccaf
â”‚Â Â  â”‚Â Â  â”‚Â Â  â””â”€â”€ validate_resolution
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ cluster.number_neighbors=-1.method=leiden.resolution=0.5
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ cluster.number_neighbors=-1.method=leiden.resolution=1.0
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ cluster.number_neighbors=-1.method=leiden.resolution=5.0
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ outfile_adatabbknn.h5ad
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ plots
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ reduced_dims.tsv.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ resolution_tuningmerged_model_report.tsv.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ resolution_tuningmerged_test_result.tsv.gz
â”‚Â Â  â”‚Â Â  â””â”€â”€ umap_gather_out.h5ad
â”‚Â Â  â”œâ”€â”€ reduced_dims-null-harmony.n_pcs=20.variables=experiment_id.thetas=1.0
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ cluster.number_neighbors=15.method=leiden.resolution=0.1
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ cluster.number_neighbors=15.method=leiden.resolution=0.5
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ cluster.number_neighbors=15.method=leiden.resolution=1.0
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ cluster.number_neighbors=15.method=leiden.resolution=5.0
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ plots
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ reduced_dims.tsv.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ resolution_tuningmerged_model_report.tsv.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ resolution_tuningmerged_test_result.tsv.gz
â”‚Â Â  â”‚Â Â  â””â”€â”€ umap_gather_out.h5ad
â”‚Â Â  â””â”€â”€ reduced_dims-null-pca.n_pcs=20
â”‚Â Â      â”œâ”€â”€ clustering_and_integration
â”‚Â Â      â”‚Â Â  â””â”€â”€ plots
â”‚Â Â      â””â”€â”€ reduced_dims.tsv.gz
â””â”€â”€ plots

```
</details>

## citeseq

The `citeseq` folder contains the following subdirectories:

- **[DSB]**  
  DSB has folders for each pool which contain DSB background and removed protein counts if citeseq is in the quantification matrix

<details>
<summary>Click to view detailed folder structure</summary>

```
citeseq/
â””â”€â”€ DSB
    â””â”€â”€ Pool1
        â””â”€â”€ CITE__Pool1
```
</details>

## handover
The `handover` folder contains the following subdirectories:
- **Summary_plots**
- **Donor_Quantification**
- **Donor_Quantification_summary**
- **merged_h5ad**
- **UMAPs**

<details>
<summary>Click to view detailed folder structure</summary>

```
handover/
â”œâ”€â”€ Donor_Quantification
â”œâ”€â”€ Donor_Quantification_summary
â”œâ”€â”€ merged_h5ad
â””â”€â”€ Summary_plots
â””â”€â”€ UMAPs
```
</details>

